### ISSUE TITLE
MissingJavadocMethod does not give warning for missing javadoc for protected methods in google_checks.xml

### ISSUE THREAD

**Issue by Zopsss** [checkstyle/checkstyle#15434]
I have read check documentation: https://checkstyle.org/checks/javadoc/missingjavadocmethod.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

Detected at: https://github.com/checkstyle/checkstyle/pull/14901#discussion_r1700067279

[7.3 Where Javadoc is used](https://google.github.io/styleguide/javaguide.html#s7.3-javadoc-where-required) rule of google java style guide states that:

> At the minimum, Javadoc is present for every public class, and every public or protected member of such a class

for [this rule](https://checkstyle.org/google_style.html#a7.3.1), we use 2 modules: [MissingJavadocMethod](https://checkstyle.org/checks/javadoc/missingjavadocmethod.html#MissingJavadocMethod)  & [JavadocMethod](https://checkstyle.org/checks/javadoc/javadocmethod.html#JavadocMethod)

MissingJavadocMethod is used to check for missing javadoc comments for methods. But this module does not give warning for missing javadoc for protected methods, example:

```java
// testing
public class JavadocTesting {

  // testing
  protected void foo() { // false negative, should give missing javadoc warning.
    int x = 99;
    System.out.println("testing");
    System.out.println(x);
    System.out.println(x);
  }

  /* testing */
  public void foo2() { // ok, warning expected here
    int x = 99;
    System.out.println("testing");
    System.out.println(x);
  }

  void foo3() {
    int x = 99;
    System.out.println("testing");
    System.out.println(x);
  }

  private void foo4() {
    int x = 99;
    System.out.println("testing");
    System.out.println(x);
  }

  // testing
  protected int xyz = 10;

  /* testing */
  public int pqr = 10;

  int abc = 10;

  private int mnp = 10;

  /* testing */
  public class Inner {

    // testing
    protected void foo() { // false negative, should give missing javadoc warning.
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
      System.out.println(x);
    }

    /* testing */
    public void foo2() { // ok, warning expected here
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
    }

    void foo3() {
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
    }

    private void foo4() {
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
    }

    // testing
    protected int xyz = 10;

    /* testing */
    public int pqr = 10;

    int abc = 10;

    private int mnp = 10;
  }

  // testing
  protected class Inner2 {

    // testing
    protected void foo() {
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
      System.out.println(x);
    }

    /* testing */
    public void foo2() {
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
    }

    void foo3() {
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
    }

    private void foo4() {
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
    }

    // testing
    protected int xyz = 10;

    /* testing */
    public int pqr = 10;

    int abc = 10;

    private int mnp = 10;
  }

  class Inner3 {

    // testing
    protected void foo() {
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
      System.out.println(x);
    }

    /* testing */
    public void foo2() {
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
    }

    void foo3() {
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
    }

    private void foo4() {
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
    }

    // testing
    protected int xyz = 10;

    /* testing */
    public int pqr = 10;

    int abc = 10;

    private int mnp = 10;
  }

  private class Inner4 {

    // testing
    protected void foo() {
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
      System.out.println(x);
    }

    /* testing */
    public void foo2() {
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
    }

    void foo3() {
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
    }

    private void foo4() {
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
    }

    // testing
    protected int xyz = 10;

    /* testing */
    public int pqr = 10;

    int abc = 10;

    private int mnp = 10;
  }
}

```

**Output:**

```
$ java -jar .\checkstyle-10.17.0-all.jar -c .\google_checks.xml JavadocTesting.java
Starting audit...
[WARN] C:\checkstyle testing\TestingGoogleStyle\JavadocTesting.java:2:1: Missing a Javadoc comment. [MissingJavadocType]
[WARN] C:\checkstyle testing\TestingGoogleStyle\JavadocTesting.java:13:3: Missing a Javadoc comment. [MissingJavadocMethod]
[WARN] C:\checkstyle testing\TestingGoogleStyle\JavadocTesting.java:42:3: Missing a Javadoc comment. [MissingJavadocType]
[WARN] C:\checkstyle testing\TestingGoogleStyle\JavadocTesting.java:53:5: Missing a Javadoc comment. [MissingJavadocMethod]
[WARN] C:\checkstyle testing\TestingGoogleStyle\JavadocTesting.java:83:3: Missing a Javadoc comment. [MissingJavadocType]
Audit done.
```

All warnings are correct but we're missing warnings for following method of public classes ( JavadocTesting, Inner ):

```java
  // testing
  protected void foo() { // false negative, should give missing javadoc warning.
    int x = 99;
    System.out.println("testing");
    System.out.println(x);
    System.out.println(x);
  }
```

The output should be something like:

```
java -jar .\checkstyle-10.17.0-all.jar -c .\google_checks.xml .\JavadocTesting.java
Starting audit...
[WARN] C:\checkstyle testing\TestingGoogleStyle\.\JavadocTesting.java:2:1: Missing a Javadoc comment. [MissingJavadocType]
[WARN] C:\checkstyle testing\TestingGoogleStyle\.\JavadocTesting.java:5:3: Missing a Javadoc comment. [MissingJavadocType]
[WARN] C:\checkstyle testing\TestingGoogleStyle\.\JavadocTesting.java:8:3: Missing a Javadoc comment. [MissingJavadocType]
[WARN] C:\checkstyle testing\TestingGoogleStyle\.\JavadocTesting.java:15:3: Missing a Javadoc comment. [MissingJavadocMethod]
[WARN] C:\checkstyle testing\TestingGoogleStyle\.\JavadocTesting.java:23:3: Missing a Javadoc comment. [MissingJavadocMethod]
[WARN] C:\checkstyle testing\TestingGoogleStyle\.\JavadocTesting.java:45:3: Missing a Javadoc comment. [MissingJavadocMethod]
Audit done.
```

There's a defect in our google config for this module:

https://github.com/checkstyle/checkstyle/blob/6a3ab019020cf67973e4f3fdc8ace6d197eadddc/src/main/resources/google_checks.xml#L339-L344

The `scope` property is used to define which modifiers should be considered for this module, so the defect is in this property. 

---

**Describe what you expect in detail.**

Fix the `scope` property to also consider "protected" methods for this module, so we can detect such methods, give warning for them and follow the rule entirely. 

---

**Doubt:**: I have a doubt, do we need to give missing javadoc warning for public & protected variables also? If yes, then there's another defect in our config as we don't give violations for them

**Comment by rnveach** [comment]
If valid, it looks like a scope issue and not a check issue. 

Issue does not mention this is connected to Google style and does not list the section that says style demands this.

**Comment by romani** [comment]
Issue needs to pass mentors review and approval... I will review it a bit later.

**Comment by Zopsss** [comment]
> If valid, it looks like a scope issue and not a check issue.
> 
> Issue does not mention this is connected to Google style and does not list the section that says style demands this.

@rnveach I have updated the issue description, referenced the google documentation and explained the issue more clearly.

**Comment by rnveach** [comment]
>  Javadoc is present for every public class, and every public or protected member of such a class

This sounds like it also means private classes with public/protected members are excluded. I am not sure the check provides this fine grain control.

**Comment by Zopsss** [comment]
yes private classes are excluded, only public classes are considered, specifically only their public and protected members are considered. I have updated the example in the description, explaining where warnings should appear.

**Comment by Zopsss** [comment]
@romani I was trying to check that how to detect protected methods for public classes. 

In MissingJavadocType, `scope` is set to `protected` to detect missing javadoc for public and protected classes:

https://github.com/checkstyle/checkstyle/blob/1f85c8e9860b24235536b41ca9605c69945d3af4/src/main/resources/google_checks.xml#L346-L352

In MissingJavadocMethod, `scope` is set to `public` to detect missing javadoc for public methods:

https://github.com/checkstyle/checkstyle/blob/1f85c8e9860b24235536b41ca9605c69945d3af4/src/main/resources/google_checks.xml#L339-L345

if I change MissingJavadocMethod's `scope` to protected then protected methods of a protected class also gets detected which we does not want. Can you guide me with how we should approach this?

```java
public class TestingPublicProteted { // line 1
  public void foo() { // line 2
    int x = 99;
    System.out.println("testing");
    System.out.println(x);
    System.out.println(x);
  }

  protected void foo2() { // line 9
    int x = 99;
    System.out.println("testing");
    System.out.println(x);
    System.out.println(x);
  }

  protected class Inner { // line 16
    public void foo() { // line 17
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
      System.out.println(x);
    }

    protected void foo2() { // line 24
      int x = 99;
      System.out.println("testing");
      System.out.println(x);
      System.out.println(x);
    }
  }
}
```

**Modified part of config:**
```xml
    <module name="MissingJavadocMethod">
      <property name="scope" value="protected"/>
      <property name="minLineCount" value="2"/>
      <property name="allowedAnnotations" value="Override, Test"/>
      <property name="tokens" value="METHOD_DEF, CTOR_DEF, ANNOTATION_FIELD_DEF,
                                   COMPACT_CTOR_DEF"/>
    </module>
    <module name="MissingJavadocType">
      <property name="scope" value="protected"/>
      <property name="tokens"
                value="CLASS_DEF, INTERFACE_DEF, ENUM_DEF,
                      RECORD_DEF, ANNOTATION_DEF"/>
      <property name="excludeScope" value="nothing"/>
    </module>
```

**Output of original config:**
```
$ java -jar .\checkstyle-10.17.0-all.jar -c .\google_checks.xml .\TestingPublicProteted.java
Starting audit...
[WARN] .\TestingPublicProteted.java:1:1: Missing a Javadoc comment. [MissingJavadocType]
[WARN] .\TestingPublicProteted.java:2:3: Missing a Javadoc comment. [MissingJavadocMethod]
[WARN] .\TestingPublicProteted.java:16:3: Missing a Javadoc comment. [MissingJavadocType]
Audit done.
```

**Output of modified config:**
```
$ java -jar .\checkstyle-10.17.0-all.jar -c .\google_checks.xml .\TestingPublicProteted.java
Starting audit...
[WARN] .\TestingPublicProteted.java:1:1: Missing a Javadoc comment. [MissingJavadocType]
[WARN] .\TestingPublicProteted.java:2:3: Missing a Javadoc comment. [MissingJavadocMethod]
[WARN] .\TestingPublicProteted.java:9:3: Missing a Javadoc comment. [MissingJavadocMethod]
[WARN] .\TestingPublicProteted.java:16:3: Missing a Javadoc comment. [MissingJavadocType]
[WARN] .\TestingPublicProteted.java:17:5: Missing a Javadoc comment. [MissingJavadocMethod]
[WARN] .\TestingPublicProteted.java:24:5: Missing a Javadoc comment. [MissingJavadocMethod]
Audit done.
```

**Comment by romani** [comment]
@Zopsss

> Doubt:: I have a doubt, do we need to give missing javadoc warning for public & protected variables also? If yes, then there's another defect in our config as we don't give violations for them

please create separate issue on this. We need to create new Check or reuse https://checkstyle.org/checks/javadoc/javadocvariable.html

**Comment by romani** [comment]
@Zopsss

>This sounds like it also means private classes with public/protected members are excluded. I am not sure the check provides this fine grain control.

please send PR with Inputs that shows how it behaves now on such edge-case.

**Comment by romani** [comment]
>if I change MissingJavadocMethod's scope to protected then protected methods of a protected class also gets detected which we does not want. Can you guide me with how we should approach this?

we do need violations on them too. They are accesible from other classes so should be documented, this is whole idea of javadocs - document what is possible to use.
Please create PRs with Inputs to see edge-cases in real execution.

---------

I am aprooving this issue to start updates and see that edge-cases we should not violate, we will use xpath suppressions or ...... to avoid them.
